[[ Bug 18998 ]] BOM Script-only Stack

bwmilby opened the pull request:
Handle BOM for Script-only stacks from URLs
Testing code to identify where the issue can be handled
Need to re-work to strip BOM and pass to ScriptPoint.

bwmilby committed:
[[ Bug 18998 ]] [WIP] BOM

Handle BOM for Script-only stacks from URLs
Testing code to identify where the issue can be handled

montegoulding commented:
@bwmilby there's a bit of an issue here in that I believe script stacks with a BOM for other encodings are able to be read by the engine... 

@runrevmark I'm not really sure why we don't throw an error if there's some other BOM when readingnif we _always_ write UTF8 with BOM. Also no BOM should probably still assume UTF8 given that's all the engine writes out. That way we have two cases. UTF8 with a BOM and UTF8 without one. We could also then change it so the engine does not write a BOM as I have noticed this is a constant source of diffs in our patches as obviously some people are editing with text editors that do not write it.

bwmilby commented:
Looking at the code, opening a local file should work with UTF16 BE/LE and a couple others too.  I was able to capture a file with a UTF8 BOM, so I used that for testing.  I placed it on my server with and without the BOM to test pulling in using the URL syntax.  My code was mainly to prove to myself that I had identified where the test was failing.  I didn't realize that files got saved out with the UTF8 BOM (even if opened without one).

bwmilby committed:
Use MCInterfaceTryToEvalStackFromString directly

The code that opens a stack from the file system uses MCInterfaceTryToEvalStackFromString.  Since the input is a MCStringRef, it would be a drop-in replacement.

If that call failed to retrieve a stack, then MCInterfaceTryToEvalStackFromString should fail too.  I'm not sure if it would be proper to directly set the error message after the MCInterfaceStringCouldBeStack test.  Worst case, the file is evaluated a second time before the error is returned with the nil pointer.

bwmilby committed:
correct comment typo

bwmilby commented:
Seeing that this change doesn't show errors in the message box, however that is the same in the latest DP.

bwmilby committed:
Add MCresult->clear() if it was not a stack and did not look like a binary stack in the header.  Also, no need to eval the file again.

bwmilby commented on engine/src/cmdss.cpp:
```diff
@@ -448,16 +448,22 @@ Parse_stat MCGo::parse(MCScriptPoint &sp)
+        MCresult->clear();
     
 	if (etype == CT_STACK)
         return NULL;
```
Why is nil returned for some cases and NULL in others?  There are 2 calls in MCGo::exec_ctxt to findstack.  The first uses (sptr==nil) and the second uses (sptr == NULL) to test for success.  My understanding is that the difference is the type (nullptr vice int).  In foundation.h there is a define for nil (C++ compiler) to nullptr.

livecodeali replied:
That's a relic of old code. For a while we started to use `nil` by preference over `NULL`. Now, in new code, `nullptr` is preferred. Rather than do any kind of global replace (which is a bit of a nightmare for merging and so on), we just change them when the relevant code is touched.

mwieder replied:
@livecodeali Good to know. Thanks. I'll use nullptr from now on.

montegoulding replied:
I didn't know we had a preference either @mwieder ... actually I thought the preference was `nil`.. Anyway, I will also use `nullptr` from now on. Actually having looked again I see it is mentioned in initialising local variables here https://github.com/livecode/livecode/blob/develop/docs/development/C%2B%2B-style.md

bwmilby committed:
Update nil/NULL to nullptr in findstack and the 2 uses of the return value
Also converted tabs to spaces around lines already being edited

bwmilby committed:
Reverting to code closer to the original state.
If a BOM is present in p_value, then the text should have come from a URL download.  In that case, the only way that it could be processed here would be as a script-only stack (the other options in findstack don't make much sense, i.e. findstackname_string).  Due to that, it should be fine to run anything with a UTF-8 BOM through the MCInterfaceTryToEvalStackFromString.

runrevmark commented on engine/src/exec-interface2.cpp:
```diff
@@ -2480,8 +2480,18 @@ bool MCInterfaceStringCouldBeStack(MCStringRef p_string)
-    
-    // Check if it could be a script-only stack
+
+    //TODO: process BOM at the point of import into LiveCode (from URL)
```
I'm not sure this is necessary given the way this PR works - it wouldn't really add anything, and is actually quite tricky given that (at present) we basically treat the StringRef as Data in several of these functions (which is correct because of the way Data<->String conversions work, just a little bit murky).

bwmilby replied:
My thought was that if a URL keyword was presented, then the developer probably intended for the result to be decoded text (and the idea is that BOMs don't occur in usual binary data streams), so if processing the URL keyword a BOM check and decode based on that would be useful.  In any case, I'll remove the comment since that would be a separate issue to track/change.

runrevmark commented on engine/src/exec-interface2.cpp:
```diff
@@ -2480,8 +2480,18 @@ bool MCInterfaceStringCouldBeStack(MCStringRef p_string)
+    //UTF-8 BOM would most likely result from a URL chunk.  It would
+    //not be evaluated any other way, so return true to have it evaluated
+    //as a script-only stack.
+    const char_t t_UTF8BOM[] = { 0xEF, 0xBB, 0xBF, 0x00 };
```
It might be worth justifying the 12 here in a comment. i.e.

Here we check whether it is very likely that the string is a UTF-8 encoded script-only stack and not a name. Any UTF-8 encoded script-only stack with BOM will start with 'BOMscript "'. - i.e. have the 3 byte BOM prefix and then 'script "'. 

And tightening the check. i.e. You can make the string to look for: { 0xEF, 0xBB, 0xBF, 's', 'c', 'r', 'i', 'p', 't', ' ', '"', 0x00 } which tightens up the check as far as it can be.

bwmilby replied:
My first attempt actually did that.  I backed it out since the parser did allow blank lines before script in another place.  I'll make these adjustments.

runrevmark replied:
I think we can tighten that up - and make it much more strict everywhere - thus making this case easier and ensuring there isn't too much variance in stack headers (which makes all tooling easier).

bwmilby replied:
script_only_stack_from_bytes would be a place to include the check.  At that point, the BOM is removed so it would be easy to check for "script" as the first word and return nullptr if not.

runrevmark commented on engine/src/exec-interface2.cpp:
```diff
@@ -2480,8 +2480,18 @@ bool MCInterfaceStringCouldBeStack(MCStringRef p_string)
+    //UTF-8 BOM would most likely result from a URL chunk.  It would
+    //not be evaluated any other way, so return true to have it evaluated
+    //as a script-only stack.
+    const char_t t_UTF8BOM[] = { 0xEF, 0xBB, 0xBF, 0x00 };
```
You could make the string to look for: { 0xEF, 0xBB, 0xBF, 's', 'c', 'r', 'i', 'p', 't', ' ', '"', 0x00 } which tightens up the check as far as it can be.

runrevmark commented:
@bwmilby : This definitely looks like the correct way to go. The only thing we might want to do (in addition) is submit an issue (to be done as a separate PR - as it is orthogonal) which constrains the structure of the first line of a script-only stack - i.e. make it so that it has to be strictly 'script \"' - i.e. no space before the script, and only one space between script and \".

bwmilby committed:
Add check for "script" after UTF-8 BOM

bwmilby committed:
Add { ' ', '"' } to the check (missed that in the first update)

bwmilby commented:
I guess I shouldn't commit code first thing in the morning.  I'll fix this evening.

bwmilby committed:
Update comment, correct typo in t_UTF8BOM declaration.

runrevmark commented on engine/src/exec-interface2.cpp:
```diff
@@ -2480,8 +2480,18 @@ bool MCInterfaceStringCouldBeStack(MCStringRef p_string)
+    //encoded script-only stack and not a name. Any UTF-8 encoded script-only
+    //stack with BOM will start with 'BOMscript "'. - i.e. have the 3 byte
+    //BOM prefix and then 'script "'.
+    const char_t t_UTF8BOM[] = { 0xEF, 0xBB, 0xBF, 's', 'c', 'r', 'i', 'p',
```
I suggest perhaps using k_ prefix rather than t_ here - as it is a constant (I'm not sure we cover const .. type vars in our style guide at the moment). Also, generally lower-case var names are preferred - k_utf8_bom.

bwmilby committed:
Update t_UTF8BOM to k_utf8bom

runrevmark commented:
@livecode-vulcan review ok 745af0d

livecode-vulcan commented:
💙 review by @runrevmark ok 745af0db5498546e34531dd4ea7a32b0bb55ccee

livecode-vulcan commented:
😞 test failure 745af0db5498546e34531dd4ea7a32b0bb55ccee

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/616)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/632)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/606)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/549)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/552)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/610)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/630)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/615)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/632)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/614)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/625)
* try-community-x86-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/647)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/611)

runrevmark commented:
@livecode-vulcan retry 745af0d

livecode-vulcan commented:
😎 test success 745af0db5498546e34531dd4ea7a32b0bb55ccee

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/618)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/634)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/608)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/551)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/554)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/612)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/632)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/617)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/634)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/616)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/627)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/649)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/613)

runrevmark commented:
@bwmilby The code is all fine - could you add a test or two for this before we merge?

bwmilby commented:
I started a test, but am having problems getting it to pass.

runrevmark commented:
@bwmilby If you push the test you've written then we're more than happy to take a look :)

bwmilby committed:
[WIP] Test for BOM import.
Test does not pass on my system yet.  Will look more this evening and see if I can trace what I'm missing.

montegoulding commented on tests/lcs/core/files/bom.livecodescript:
```diff
@@ -0,0 +1,38 @@
+   
+   // Technically this is testing the opening of a stack that
+   // is coming from a chunk.  Most likely source would be URL
+   // "binfile:" and "http(s):", but others could be possible.
```
Yes, people have been known to stuff stacks into custom properties. It would be good to add a test without the BOM here too. BTW GitHub diff is not showing me the BOM on the file. I can't recall if it normally does or not.

bwmilby replied:
The file should be good.  It does report the size as 62 bytes on GitHub which matches my local system copy.  In DP8 I can open it using binfile only with the char 4 to -1 trick.  Now that I'm thinking about it though, maybe I just need to create a variable with the BOM and use that.

montegoulding replied:
You could create the whole stack in a variable and then save it out to do `go from url` tests and use the variable directly for `go` tests. In both cases with and without BOM.

bwmilby replied:
I could change over to writing a file out and then opening it if this method isn't desired.  I did add tests that use a constructed variable for the stack (with and without a BOM).

bwmilby committed:
Second test passes, not sure why the first is failing though.

montegoulding commented on tests/lcs/core/files/bom.livecodescript:
```diff
@@ -21,6 +21,7 @@ constant kTestFile = "utf8bom.livecodescript"
    local tUrlBinfilePath
+   local tScriptOnlyStack
    
    put quote & "binfile:" & testFilesFolder() & slash & \
```
you don't need the `quote &` here or the other end

bwmilby replied:
That was the problem.  Test works now.  Will clean up and post.

bwmilby committed:
Corrected error in test file.  Added no-BOM test.

montegoulding commented:
@livecode-vulcan review ok fdcaa2b

Note to merger please squash merge to kill the WIP commits

livecode-vulcan commented:
💙 review by @montegoulding ok fdcaa2b2de7bd792cbe1ee53edd63254a6adcce5

livecode-vulcan commented:
😎 test success fdcaa2b2de7bd792cbe1ee53edd63254a6adcce5

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/673)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/689)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/663)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/606)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/609)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/667)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/687)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/672)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/689)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/671)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/682)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/704)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/668)
